typing/typecore.ml

```diff
@@ -4018,7 +4172,22 @@ and type_expect_
         eexp
   | None      -> match sexp.pexp_desc with
   | Pexp_ident lid ->
-      let path, mode, desc, kind = type_ident env ~recarg lid in
+      let path, mode, shared_reason, desc, kind = type_ident env ~recarg lid in
+      let mode =
+        match path with
+        | Path.Pident _ ->
+            let uniqueness, _ =
+              Uniqueness.newvar_above
+                (Value.uniqueness mode)
```

Comment:
That makes sense but I think there is an issue: `mode` retreived from the environment is a "left" value, but later we might want to say `shared <= mode` if we want to enforce it. Indeed, this will be not even typable once the new solver kicks in.

Reply:
Actually, I think this whole part is needlessly confusing. Fundamentally, we place two kinds of restrictions on modes from the uniqueness analysis:
1. We constrain the variable to be used `shared` -- that means we need a "right" uniqueness mode.
2. We constraint the variable to be defined `many` -- that means we need a "left" linearity mode.
We already have such modes available without creating any variables: we can use the uniqueness of the current expected mode, and the linearity that comes from the environment.

In the case that the extension is disabled we can just constrain those two modes to be `shared` and `many` respectively. This also avoids the need to distinguish the `Path.Pident` case since we won't be creating any extra variables.

Reply:
Yeah that makes perfect sense.